docs/tsp/typeServerProtocol.ts

```diff
@@ -0,0 +1,383 @@
+        [Requests.GetSymbolDeclarationInfo]: {
+            node: Node;
+            skipUnreachableCode: boolean;
+            snapshot: number;
+        };
+        [Requests.GetBoundMagicMethod]: { type: Type; name: string; snapshot: number };
+        [Requests.ResolveAliasDeclaration]: {
+            decl: Declaration;
+            options: ResolveAliasOptions;
+            snapshot: number;
+        };
+        [Requests.ResolveImport]: ResolveImportParams;
+        [Notifications.Initialized]: void;
+        [Notifications.Shutdown]: void;
+        [Notifications.FlushCache]: void;
+        [Notifications.TextDocumentOpen]: TextDocumentOpenParams;
+        [Notifications.TextDocumentClose]: TextDocumentCloseParams;
+        [Notifications.SettingsChange]: Settings;
+        [Notifications.FilesChanged]: { changes: FileEvent[] };
+    }
+
+    export interface Response {
+        [Requests.Initialize]: { configRoots: string[] };
+        [Requests.Shutdown]: { success: boolean };
+        [Requests.GetSnapshot]: number;
+        [Requests.GetDiagnostics]: Diagnostic[] | undefined;
+        [Requests.GetDiagnosticsVersion]: number | undefined;
+        [Requests.GetType]: Type | undefined;
+        [Requests.GetUnionTypes]: Type[] | undefined;
+        [Requests.GetTypeMember]: Member | undefined;
+        [Requests.GetOverloads]: Type[] | undefined;
+        [Requests.GetEffectiveMetaClass]: Type | undefined;
+        [Requests.GetTypeOfDeclaration]: Type | undefined;
+        [Requests.GetSymbolDeclarationInfo]: SymbolDeclInfo | undefined;
+        [Requests.GetBoundMagicMethod]: Type | undefined;
+        [Requests.ResolveAliasDeclaration]: Declaration | undefined;
+        [Requests.ResolveImport]: string | undefined;
+    }
+
+    export function sendRequest<P extends Params, R extends Response, M extends Requests & keyof P & keyof R & string>(
+        connection: RequestSender,
+        method: M,
+        params: P[M],
+        token?: CancellationToken
+    ): Promise<R[M]> {
+        return connection.sendRequest(method, params, token);
+    }
+
+    export function sendNotification<P extends Params, M extends Notifications & keyof P & string>(
+        connection: NotificationSender,
+        method: M,
+        params: P[M]
+    ): void {
+        connection.sendNotification(method, params);
+    }
+
+    export function onRequest<P extends Params, R extends Response, M extends Requests & keyof P & keyof R & string, E>(
+        connection: RequestReceiver,
+        method: M,
+        handler: RequestHandler<P[M], R[M], E>
+    ): Disposable {
+        return connection.onRequest(method, handler);
+    }
+
+    export function onNotification<P extends Params, M extends Notifications & keyof P & string>(
+        connection: NotificationReceiver,
+        method: M,
+        handler: NotificationHandler<P[M]>
+    ): Disposable {
+        return connection.onNotification(method, handler);
+    }
```

Comment:
@erictraut, thanks a lot for all of your feedback. Internally I'm going to try and make a lot of the changes you suggested and see how it affects the semantic token provider.  I'll include you on the review list on the next update to the public version.